vowpalwabbit/array_parameters_dense.h

```diff
@@ -47,10 +49,19 @@ class dense_iterator
 class dense_parameters
 {
 private:
+  // struct to store the tag hash and if it is set or not
+  struct tag_hash_info
+  {
+    uint64_t tag_hash;
+    bool is_set = false;
+  };
   weight* _begin;
   uint64_t _weight_mask;  // (stride*(1 << num_bits) -1)
   uint32_t _stride_shift;
   bool _seeded;  // whether the instance is sharing model state with others
+  size_t _privacy_activation_threshold;
+  std::unordered_map<uint64_t, std::bitset<32>> _feature_bitset;  // define the bitset for each feature
```

Comment:
this could be a unique_ptr set to nullptr unless the privacy mode is on, to be a bit explicit and avoid extra memory allocations

Reply:
I like that suggestion, making it a shared ptr due to shallow copy fn

Reply:
done